CodegenProposal.md

```diff
@@ -0,0 +1,1474 @@
+
+For this reason, we have opted to not generate default values.
+
+## Concrete Subtypes as Enums
+
+[@designatedNerd’s initial proposal includes enums with associated values for subtypes.](https://github.com/apollographql/StarWarsiOSMk2/blob/master/StarWarsMark2/Queries/HeroTypeDependentAliasedFieldQueryMk2.swift#L43)
+
+These `SubType` enums thing works for concrete types, not interfaces (because a type could conform to multiple interfaces). We don't plan on generating all the Concrete types as data structures unless they are specifically enumerated (`... on Droid`)
+
+It is undecided if we should implement these or not. They are only valuable in the specific scenario where you have inline fragments for multiple concrete types. 
+
+We could easily generate the `asDroid: Droid?` and `asHuman: Human?`  properties on the object.
+
+Given this query:
+
+```graphql
+query {
+    hero {
+         ... on Human {
+            name
+        }
+        ... on Droid {
+            primaryFunction
+        }
+    }
+}
+```
+
+Without the `Subtypes` enum:
+
+```swift
+struct Hero: RootSelectionSet {
+  var asHuman: AsHuman? { asType() }
+  var asDroid: AsDroid? { asType() }
+  
+  struct AsHuman { ... }
+  struct AsDroid { ... }
+}
+```
+
+With the `Subtypes` enum:
+
+```swift
+struct Hero: RootSelectionSet {
+  var asHuman: AsHuman? { asType() }
+  var asDroid: AsDroid? { asType() }
+  
+  enum Subtype {
+    case human(AsHuman)
+    case droid(AsDroid)
+    case _other(Hero)
+  }
+
+  var subtype: Subtype {
+    switch __objectType {
+    case is Human.Type: return .human(AsHuman(data: data))
+    case is Droid.Type: return .droid(AsDroid(data: data))    
+    default: return ._other(self)
+    }
+  }
+
+  struct Human { ... }
+  struct Droid { ... }
+}
+```
+
+Possible Options:
+
+* Don't implement the subtypes enum at all
+* Use a directive `@generateSubTypeEnum` (or some other name) to inform us that the subtypes enum should be generated (if the directive is not present default is option #1.)    
+    ```graphql
+    query {
+      hero @generateSubTypeEnum {
+        ... on Human {
+          name
+        }
+        ... on Droid {
+          primaryFunction
+        }
+      }
+    }
+    ```
+* Implement logic so that if your query has _**one or more**_ fragments on a concrete type, then we generate the subtypes (generate the enum with only 1 case + `_other`)
+* Implement logic so that if your query has _**more than one**_ fragment on a concrete type, then we generate the subtypes
+
+# Possible Future Additions
+
+## Client-Side Directives For Automatic `CacheKeyProvider` Generation
```

Comment:
Isn't this covered by the cache key stuff abovve?

Reply:
It's mentioned briefly above. I just wanted to have a specific section that we compile a list of possible future considerations.

Reply:
And this isn't being implemented right now. The stuff above is examples of how you can implement `CacheKeyProviders`, but they aren't going to be generated for you. You have to write them manually. 

This refers to using the `@keyFields` directive that @martinbonnin has been working on defining to generate these for you automatically.

Reply:
Gooot it. Might be worth filling in brief details on what these things are before we call in the public.

Reply:
Done